opencog/reasoning/RuleEngine/rules/pln/evaluation-to-member.scm

```diff
@@ -0,0 +1,58 @@
+; ============================================================================= 
+; GeneralEvaluationToMemberRule 
+;	Takes EvaluationLinks with >=2 Arguments and creates a Member Link
+;
+;	EvaluationLink (pred D (ListLink B C))
+;		becomes
+;	MemberLink( B (SatisfyingSetLink( Variable $X 
+;		(EvaluationLink (pred D (ListLink X C))))))
+; -----------------------------------------------------------------------------
+
+(define pln-rule-evaluation-to-member
+	(BindLink
+		(ListLink
+			(VariableNode "$A")
+			(TypedVariableLink
+				(VariableNode "$D")
+				(VariableTypeNode "PredicateNode")))
+        (ImplicationLink
+            (EvaluationLink
+                (VariableNode "$D")
+                (VariableNode "$A"))
+            (ExecutionOutputLink
+                (GroundedSchemaNode "scm:pln-formula-evaluation-to-member")
+                (ListLink
+                    (EvaluationLink
+                        (VariableNode "$D")
+                        (VariableNode "$A")))))))
+
+
+; -----------------------------------------------------------------------------
+; Evaluation To Member Formula
+; -----------------------------------------------------------------------------
+
+; -----------------------------------------------------------------------------
+; Side-effect: TruthValue of the new link stays the same
+; -----------------------------------------------------------------------------
+
+(define (pln-formula-evaluation-to-member DBC)
+	(if (= (cog-arity (gdr DBC)) 0)
+		(MemberLink (stv (cog-stv-strength DBC) (cog-stv-confidence DBC))
+			(gdr DBC)
+			(SatisfyingSetLink
+				(VariableNode "$X")
+				(EvaluationLink
+					(gar DBC)
+					(VariableNode "$X"))))
+		(MemberLink (stv (cog-stv-strength DBC) (cog-stv-confidence DBC))
+			(gadr DBC)
+			(SatisfyingSetLink
+				(VariableNode "$X")
+				(EvaluationLink
+					(gar DBC)
+					(ListLink
+						(cons
+							(VariableNode "$X")
+							(cdr (cog-get-all-nodes (gdr DBC))))))))))
```

Comment:
I think replacing ```cog-get-all-nodes``` with ```cog-outgoing-set``` should be able to handle links as arguments of predicate. Also, I am not sure about the expected output of EvaluationLink generating MemberLink for each argument.
For Example:-
```
(EvaluationLink (stv 0.9 0.9)
    (PredicateNode "give")
    (ListLink
        (ConceptNode "Ram")
        (ConceptNode "Rahim")
        (ConceptNode "Book")))
```
should give us
```
(MemberLink
    (ConceptNode "Ram")
    (SatisfyingSetLink
        (VariableNode "$X")
        (EvaluationLink
            (PredicateNode "give")
            (ListLink
                (VariableNode "$X")
                (ConceptNode "Rahim")
                (ConceptNode "Book")))))
```
and
```
(MemberLink
    (ConceptNode "Rahim")
    (SatisfyingSetLink
        (VariableNode "$X")
        (EvaluationLink
            (PredicateNode "give")
            (ListLink
                (ConceptNode "Ram")
                (VariableNode "$X")
                (ConceptNode "Book")))))
```
and
```
(MemberLink
    (ConceptNode "Book")
    (SatisfyingSetLink
        (VariableNode "$X")
        (EvaluationLink
            (PredicateNode "give")
            (ListLink
                (ConceptNode "Ram")
                (ConceptNode "Rahim")
                (VariableNode "$X")))))
```
Does this look right? Also, if yes, should the TVs associated with all the links created be same or should they be divided in a particular precedence?